src/lens.jl

```diff
@@ -92,8 +100,10 @@ get(obj, ::IdentityLens) = obj
 set(obj, ::IdentityLens, val) = val
 
 struct PropertyLens{fieldname} <: Lens end
+struct PropertyLens!{fieldname} <: Lens end
+struct PropertyLens!!{fieldname} <: Lens end
```

Comment:
I would implement it using something like this:

```julia
abstract type MutationPolicy end
struct Immutable <: MutationPolicy end  # no bang
struct AlwaysMutate <: MutationPolicy end  # !
struct TryMutate <: MutationPolicy end  # !!

struct PropertyLens{fieldname, TP <: MutationPolicy} <: Lens
    policy::TP
end

PropertyLens(fieldname, policy = Immutable()) =
    PropertyLens{fieldname, typeof(policy)}(policy)
```

It would also be handy define accessor functions

```julia
MutationPolicy(::Lens) = Immutable()  # should it be this?
MutationPolicy(l::PropertyLens) = l.policy
```

Reply:
Yeah that seems more concise (and also replaces the fairly redundant `BangStyle` trait). To maintain `PropertyLens` being a Singleton, what about:
```julia
struct PropertyLens{fieldname, TP} end
PropertyLens(fieldname, policy::MutationPolicy = Immutable()) = PropertyLens{fieldname, policy}()
MutationPolicy(::PropertyLens{<:Any, TP}) = TP
```

Also, just curious, what does `TP` stand for?

Reply:
> To maintain `PropertyLens` being a Singleton

It would be a singleton if `plicy` is a singleton:

```julia
struct Demo1 end
struct Demo2{T}
    demo::T
end

julia> sizeof(Demo2(Demo1()))

julia> Base.issingletontype(typeof(Demo2(Demo1())))
true
```

If we want to _enforce_ the policy to be a singleton, I think your approach would be better.

> what does `TP` stand for?

I was thinking Type-of-Policy. It can have a better name, too :)

Reply:
Ahh, I didn't know that! I was operating off of the definition in the docs: 

> Immutable composite types with no fields are singletons

which I took to imply "singleton if-and-only-if immutable composite type with no fields". Thanks for the explanation!

Perhaps `MP` for Mutation Policy?

Reply:
Maybe? We can also call it simply `Policy`, as done in e.g. `Broadcasted`: https://github.com/JuliaLang/julia/blob/d5d5718b85d829aff6c61ab23ad2d91a50f8688b/base/broadcast.jl#L169-L173

Reply:
I like the verbosity.

Reply:
* I would rename `Immutable -> NeverMutate`. It is more consistent with the other variants and we might want the name `Immutable` for another trait.
* I think `MutationPolixy(l::Lens) = NeverMutate()` is the right default. I think `Setfield` still is mainly about manipulation of immutable structs. 
* The design should be so that an ad hoc defined lens only needs to implement `get(obj, lens)` and `set(obj, lens, val)` and still works if it is not interested in mutating things.